doc/add_datafit.rst

```diff
@@ -0,0 +1,120 @@
+:orphan:
+
+How to add a custom datafit
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _how:
+
+Motivated by generalized linear models but not limited to it, ``skglm`` solves problems of the form
+
+.. math::
+      \hat{\beta} \in
+      \arg\min_{\beta \in \mathbb{R}^p}
+      F(X\beta) + \Omega(\beta)
+      := \sum_{i=1}^n f_i(X_i^{\top}\beta) + \sum_{j=1}^p \Omega_j(\beta_j)
+      \enspace .
+
+
+Here, :math:`X \in \mathbb{R}^{n \times p}` denotes the design matrix with :math:`n` samples and :math:`p` features,
+and :math:`\beta \in \mathbb{R}^p` is the coefficient vector.
+
+``skglm`` can solve any problems of this form with arbitrary smooth datafit :math:`F` and arbitrary proximable penalty :math:`\Omega`,
+by defining two classes: a ``Penalty`` and a ``Datafit``.
+
+They can then be passed to a :class:`~skglm.GeneralizedLinearEstimator`.
+
+.. code-block:: python
+
+   clf = GeneralizedLinearEstimator(
+      MyDatafit(),
+      MyPenalty(),
+   )
+
+
+A ``Datafit`` is a jitclass which must inherit from the ``BaseDatafit`` class:
+
+.. literalinclude:: ../skglm/datafits/base.py
+   :pyobject: BaseDatafit
+
+
+To define a custom datafit, you need to implement the methods declared in the ``BaseDatafit`` class (value, gradient, gradient when the design matrix is sparse).
+As an example, we show how to implement the Poisson datafit in ``skglm``.
+
+
+A case in point: defining Poisson datafit
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, this requires deriving some quantities used by the solvers like the gradient or the Hessian matrix of the datafit.
+With :math:`y \in \mathbb{R}^n` the target vector, the Poisson datafit reads
+
+.. math::
+    F(X\beta) = \frac{1}{n}\sum_{i=1}^n \exp(X_i^{\top}\beta) - y_i X_i^{\top}\beta
+    \enspace .
+
+
+Let's define some useful quantities to simplify our computations. For :math:`z \in \mathbb{R}^n` and :math:`\beta \in \mathbb{R}^p`,
+
+.. math::
+   f(z) = \sum_{i=1}^n f_i(z_i)  \qquad  F(\beta) = f(X\beta)
+   \enspace .
+
+
+Computing the gradient of :math:`F` and its Hessian matrix yields
+
+.. math::
+   \nabla F(\beta) = X^{\top} \underbrace{\nabla f(X\beta)}_\textrm{raw grad} \qquad \nabla^2 F(\beta) = X^{\top} \underbrace{\nabla^2 f(X\beta)}_\textrm{raw hessian} X
+   \enspace .
+
+
+Besides, it directly follows that
+
+.. math::
+   \nabla f(z) = (f_i'(z_i))_{i \in [n]} \qquad \nabla^2 f(z) = \textrm{diag}(f_i''(z_i))_{i \in [n]}
```

Comment:
To my knowledge, $i \in [n]$ is not commune notation.
We better use $1 \leq i \leq n$ instead of $i \in [n]$.

Reply:
it is a common notation